### ISSUE TITLE
Danger isn't working in CI

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#5267]
I merged the Danger PR (#4988) earlier, but as @tomhughes pointed out, [it doesn't appear to be working](https://github.com/openstreetmap/openstreetmap-website/actions/runs/11370083928/job/31629028217?pr=5266):

```
Run bundle exec danger --verbose
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
To use multipart middleware with Faraday v2.0+, install `faraday-multipart` gem; note: this is used by the ManageGHES client for uploading licenses
fatal: couldn't find remote ref refs/heads/osmf_wiki_links
fatal: couldn't find remote ref refs/heads/osmf_wiki_links
fatal: couldn't find remote ref refs/heads/osmf_wiki_links
fatal: couldn't find remote ref refs/heads/osmf_wiki_links
From https://github.com/openstreetmap/openstreetmap-website
 * [new branch]          forms      -> origin/forms
 * [new branch]          next       -> origin/next
 * [new tag]             live       -> live
bundler: failed to load command: danger (/home/runner/work/openstreetmap-website/openstreetmap-website/vendor/bundle/ruby/3.1.0/bin/danger)
/home/runner/work/openstreetmap-website/openstreetmap-website/vendor/bundle/ruby/3.1.0/gems/danger-9.5.0/lib/danger/scm_source/git_repo.rb:114:in `raise_if_we_cannot_find_the_commit': Commit 44162906 doesn't exist. Are you running `danger local/pr` against the correct repository? Also this usually happens when you rebase/reset and force-pushed. (RuntimeError)
	from /home/runner/work/openstreetmap-website/openstreetmap-website/vendor/bundle/ruby/3.1.0/gems/danger-9.5.0/lib/danger/scm_source/git_repo.rb:96:in `ensure_commitish_exists_on_branch!'
	from /home/runner/work/openstreetmap-website/openstreetmap-website/vendor/bundle/ruby/3.1.0/gems/danger-9.5.0/lib/danger/request_sources/github/github.rb:129:in `setup_danger_branches'
	from /home/runner/work/openstreetmap-website/openstreetmap-website/vendor/bundle/ruby/3.1.0/gems/danger-9.5.0/lib/danger/danger_core/environment_manager.rb:61:in `ensure_danger_branches_are_setup'
...
```
I don't know what the problem is here - do you have any ideas @nenad-vujicic ?

**Comment by mmd-osm** [comment]
It seems this Danger issue might be related: https://github.com/danger/danger/issues/1103

**Comment by nenad-vujicic** [comment]
> ...
> Run bundle exec danger --verbose
> To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
> To use multipart middleware with Faraday v2.0+, install `faraday-multipart` gem; note: this is used by the ManageGHES client for uploading licenses
> fatal: couldn't find remote ref refs/heads/osmf_wiki_links
> ...
> 
> I don't know what the problem is here - do you have any ideas @nenad-vujicic ?

On my personal GitHub account, combination of "fetch-depth:0" + "creating new token" worked fine for supporting PRs from forks (also used in https://github.com/danger/danger/blob/master/.github/workflows/CI.yml). I'm not sure how it will behave on osm-website because of bots (or if OSM GitHub is registered as GitHub Enterprise account, here are more details https://danger.systems/guides/getting_started). Here are my steps:

1) Someone with OSM GitHub administrator privileges should generate token with very very carefully selected permissions

2) Replace Ln 17 of labeling.yml ("- uses: actions/checkout@v4") with 
- uses: actions/checkout@v4
  with:
    fetch-depth: 0

3) Replace Ln 22-24 of labeling.yml with something like this:
TOKEN='generated_token'
export DANGER_GITHUB_API_TOKEN=$TOKEN
export RUNNING_IN_ACTIONS=true
bundle exec danger --verbose

After these, it should start working with PRs from forked repos also (it already works fine with auto-generated PRs) if bots don't mess up things (haven't tested).

**Comment by mmd-osm** [comment]
One of the links I've posted mentioned "public_repo" scope, that's read only access to public repos.

**Comment by tomhughes** [comment]
Yes but having seen how they configure their own secret I'm not trusting them to tell me how to do it! Working on a proper solution now...

**Comment by mmd-osm** [comment]
I wouldn't use my own account to create the token, and rather create some new "OSM Danger Bot" GH account. You could also keep that token secret and reference it by variable name only.

**Comment by tomhughes** [comment]
The automatic token should be fine if we do things right - no need to configure a separate one.

**Comment by tomhughes** [comment]
Let's see if https://github.com/tomhughes/openstreetmap-website/commit/6d0c2913326fbfdf3578416853e31d7a950d97ed helps...

**Comment by mmd-osm** [comment]
I've just created a new PR, let's see how it goes.

**Comment by mmd-osm** [comment]
What about RUNNING_IN_ACTIONS=true ? This seems to be still missing.

**Comment by tomhughes** [comment]
Because it serves no purpose - nothing in the danger code ever looks at that.

**Comment by mmd-osm** [comment]
I haven’t checked the code before. They’re using it in their own Dangerfile to control junit reporting. So it’s really irrelevant for us: https://github.com/danger/danger/blob/cd913ea817a2fb9536172597303d78492a727668/Dangerfile#L56

**Comment by mmd-osm** [comment]
https://github.com/openstreetmap/openstreetmap-website/actions/runs/11407724340/job/31744367292 seems to be doing a bit better. I have replaced   `pull_request_target:     types: [opened, synchronize]` by  `on: [pull_request]`.

The token doesn't have access to remove or add labels:

```
Error message: "Big PR" label is not existing.
DELETE https://api.github.com/repos/openstreetmap/openstreetmap-website/issues/5270/labels/Big%20PR: 403 - Resource not accessible by integration // See: https://docs.github.com/rest/issues/labels#remove-a-label-from-an-issue
Error message: "Inappropriate Translations" label is not existing.
DELETE https://api.github.com/repos/openstreetmap/openstreetmap-website/issues/5270/labels/Inappropriate%20Translations: 403 - Resource not accessible by integration // See: https://docs.github.com/rest/issues/labels#remove-a-label-from-an-issue
Error message: "Merge Commits" label is not existing.
DELETE https://api.github.com/repos/openstreetmap/openstreetmap-website/issues/5270/labels/Merge%20Commits: 403 - Resource not accessible by integration // See: https://docs.github.com/rest/issues/labels#remove-a-label-from-an-issue
All green. Congrats.

Danger does not have write access to the PR to set a PR status.
```

**Comment by mmd-osm** [comment]
I found another alternative in https://github.com/chef/chef/pull/14134, which is danger-js. It seems to work on a forked repo pull request:

![image](https://github.com/user-attachments/assets/4c40858c-d32f-467f-9e54-fea1e44a6d61)

The output is a bit buried in gh action logs. Fancy things like settings tags don't seem to be supported.

![image](https://github.com/user-attachments/assets/1d7291f2-f231-45cb-8eea-f7780ef73e36)

**Comment by mmd-osm** [comment]
I think that's fine. Log files from 3 weeks ago show that danger was running back then:

![image](https://github.com/user-attachments/assets/44f417fe-9255-4e19-8092-abcb347cf55d)


https://github.com/danger/danger/actions/runs/11090746237/job/30813474684

**Comment by tomhughes** [comment]
The token is not the problem - the problem is getting it to find the commits.

**Comment by mmd-osm** [comment]
https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/ might be interesting. It describes a set up, where untrusted code is processed by an `on: pull_request` step (which has access to the pull request). In this step we could run danger, similar to what chef/chef is doing as mentioned above. The results on this analysis run can then be checked in as artifact, and another trusted CI step can then be used to download the artifact and update the Pull request labels. This second step is leveraging `on: pull_request_target:`.

I think the overall aproach might in fact work. At least the first step to run danger with  `on: pull_request` should be able to successfully analyse the untrusted code in the pull request.

**Comment by tomhughes** [comment]
I did some testing in my fork and I'm hopeful that 8e54d0f2aeaad8648a60e4685fa64380c45c5631 will actually fix it.

**Comment by mmd-osm** [comment]
It seems it's still failing: https://github.com/openstreetmap/openstreetmap-website/actions/runs/11466048440/job/31905883310?pr=5270

**Comment by tomhughes** [comment]
It worked for https://github.com/tomhughes/openstreetmap-website/actions/runs/11465903348 so it must be something specific to cross-repo PRs :-(

**Comment by mmd-osm** [comment]
Yes, it's a security feature: https://github.blog/news-insights/product-news/github-actions-improvements-for-fork-and-pull-request-workflows/

_-> new pull_request_target event [...]  runs against the workflow and code from the base of the pull request. This means the workflow is running from a trusted source and is given access to a read/write token as well as secrets enabling the maintainer to safely comment on or label a pull request._

**Comment by tomhughes** [comment]
That's not in any way relevant - it's running in the context of the base branch (master) in my repo as well but the target is still there and I'm explicitly passing it's hash as the head.

All pull_request_target means is that it doesn't merge the PR branch into master before running, so that the version of danger that runs is the master version, but the PR branch should still be present and accessible.

**Comment by mmd-osm** [comment]
I've taken my question over to https://github.com/danger/danger/issues/1103#issuecomment-2430080724

Maybe you can keep an eye the discussion over there a bit. We probably need to move to danger-js.

https://danger.systems/js/usage/danger-process as proposed in one answer seems to fit nicely to what I've written earlier today with the two separate CI steps.

**Comment by mmd-osm** [comment]
> I'm explicitly passing it's hash as the head. [...] PR branch should still be present and accessible.

I suspect that there must be more to pull_request_target to prevent untrusted code from accidentally being executed with elevated privileges...

**Comment by mmd-osm** [comment]
I'm trying the complex scenario with 2 CI steps. I've started with the second half that's updating the pull request: https://github.com/test-9bf40560-ba4d/dangertest/pull/7

Have you tried to post some comments and fail the build in case of issues?

**Comment by mmd-osm** [comment]
_I'm moving the discussion from #5290 over to this issue..._

I had some issues with danger not being able to correctly process #5080. I can also reproduce the issue in the danger test repo: https://github.com/openstreetmap/danger-test/pull/5

**Comment by mmd-osm** [comment]
Could it be that danger is checking a certain number of commits by default only? I tried adding one commit at a time in https://github.com/openstreetmap/danger-test/pull/6, and it worked up to including 15 commits. The failing PR had 20 commits.

**Comment by mmd-osm** [comment]
Danger starts failing once the PR has >= 20 commits: https://github.com/openstreetmap/danger-test/pull/6

**Comment by tomhughes** [comment]
Well that's interesting because https://github.com/danger/danger/blob/53ebd6415175ac7611b8605d5c8d20905268404c/lib/danger/scm_source/git_repo.rb#L85-L91 is the key code and that is supposed to try four passes.

The first pass looks at a depth of 20 but if that fails it should retry to 74, 222 and 625 before giving up...

**Comment by nenad-vujicic** [comment]
Can we close this now after #5295 is merged?

Thanks!

**Comment by tomhughes** [comment]
Yes I believe so.
